benches/bench.rs

```diff
@@ -39,53 +77,135 @@ pub fn jsonrpsee_types_v2(crit: &mut Criterion) {
+		let rt = TokioRuntime::new().unwrap();
+		let url = rt.block_on(helpers::ws_server());
+		let client =
+			Arc::new(rt.block_on(WsClientBuilder::default().max_concurrent_requests(1024 * 1024).build(&url)).unwrap());
+		run_round_trip(&rt, crit, client.clone(), "ws_round_trip", Self::REQUEST_TYPE);
+		run_concurrent_round_trip(&rt, crit, client, "ws_concurrent_round_trip", Self::REQUEST_TYPE);
+	}
+
+	fn batched_ws_requests(crit: &mut Criterion) {
+		let rt = TokioRuntime::new().unwrap();
+		let url = rt.block_on(helpers::ws_server());
+		let client =
+			Arc::new(rt.block_on(WsClientBuilder::default().max_concurrent_requests(1024 * 1024).build(&url)).unwrap());
+		run_round_trip_with_batch(&rt, crit, client, "ws batch requests", Self::REQUEST_TYPE);
+	}
+
+	fn subscriptions(crit: &mut Criterion) {
+		let rt = TokioRuntime::new().unwrap();
+		let url = rt.block_on(helpers::ws_server());
+		let client =
+			Arc::new(rt.block_on(WsClientBuilder::default().max_concurrent_requests(1024 * 1024).build(&url)).unwrap());
+		run_sub_round_trip(&rt, crit, client, "subscriptions");
+	}
 }
 
-pub fn websocket_requests(crit: &mut Criterion) {
-	let rt = TokioRuntime::new().unwrap();
-	let url = rt.block_on(helpers::ws_server());
-	let client =
-		Arc::new(rt.block_on(WsClientBuilder::default().max_concurrent_requests(1024 * 1024).build(&url)).unwrap());
-	run_round_trip(&rt, crit, client.clone(), "ws_round_trip");
-	run_concurrent_round_trip(&rt, crit, client, "ws_concurrent_round_trip");
+pub struct SyncBencher;
+
+impl RequestBencher for SyncBencher {
+	const REQUEST_TYPE: RequestType = RequestType::Sync;
 }
+pub struct AsyncBencher;
 
-pub fn batched_ws_requests(crit: &mut Criterion) {
-	let rt = TokioRuntime::new().unwrap();
-	let url = rt.block_on(helpers::ws_server());
-	let client =
-		Arc::new(rt.block_on(WsClientBuilder::default().max_concurrent_requests(1024 * 1024).build(&url)).unwrap());
-	run_round_trip_with_batch(&rt, crit, client, "ws batch requests");
+impl RequestBencher for AsyncBencher {
+	const REQUEST_TYPE: RequestType = RequestType::Async;
 }
 
-fn run_round_trip(rt: &TokioRuntime, crit: &mut Criterion, client: Arc<impl Client>, name: &str) {
-	crit.bench_function(name, |b| {
+fn run_round_trip(rt: &TokioRuntime, crit: &mut Criterion, client: Arc<impl Client>, name: &str, request: RequestType) {
+	crit.bench_function(&request.group_name(name), |b| {
 		b.iter(|| {
 			rt.block_on(async {
-				black_box(client.request::<String>("say_hello", JsonRpcParams::NoParams).await.unwrap());
+				black_box(client.request::<String>(request.method_name(), JsonRpcParams::NoParams).await.unwrap());
 			})
 		})
 	});
 }
 
-/// Benchmark http batch requests over batch sizes of 2, 5, 10, 50 and 100 RPCs in each batch.
-fn run_round_trip_with_batch(rt: &TokioRuntime, crit: &mut Criterion, client: Arc<impl Client>, name: &str) {
+fn run_sub_round_trip(rt: &TokioRuntime, crit: &mut Criterion, client: Arc<impl SubscriptionClient>, name: &str) {
```

Comment:
a follow is to subscribe to the "same subscription (by method name)" many times to benchmark how costly it is to access the `Mutex` when the subscriptions grows.

currently we have one `Arc<Mutex>` per registered subscription (by method name)

Reply:
I think it'd be more right to write more specific benches for that matter. Like, ones that do not include real server and client, and thus impact of mutex will be more clear.